pkg/query-service/app/logs/v4/resource_query_builder.go

```diff
@@ -0,0 +1,204 @@
+			// make sure to cast the value regardless of the actual type
+			value, err = utils.ValidateAndCastValue(item.Value, item.Key.DataType)
+			if err != nil {
+				return nil, fmt.Errorf("failed to validate and cast value for %s: %v", item.Key.Key, err)
+			}
+		}
+
+		if logsOp, ok := logOperators[op]; ok {
+			// the filter
+			if resourceFilter := buildResourceFilter(logsOp, keyName, op, value); resourceFilter != "" {
+				conditions = append(conditions, resourceFilter)
+			}
+			// the additional filter for better usage of the index
+			if resourceIndexFilter := buildResourceIndexFilter(keyName, op, value); resourceIndexFilter != "" {
+				conditions = append(conditions, resourceIndexFilter)
+			}
+		} else {
+			return nil, fmt.Errorf("unsupported operator: %s", op)
+		}
+
+	}
+
+	return conditions, nil
+}
+
+func buildResourceFiltersFromGroupBy(groupBy []v3.AttributeKey) []string {
+	var conditions []string
+
+	for _, attr := range groupBy {
+		if attr.Type != v3.AttributeKeyTypeResource {
+			continue
+		}
+		key := strings.ToLower(attr.Key)
+		conditions = append(conditions, fmt.Sprintf("(simpleJSONHas(lower(labels), '%s') AND lower(labels) like '%%%s%%')", key, key))
+	}
+	return conditions
+}
+
+func buildResourceFiltersFromAggregateAttribute(aggregateAttribute v3.AttributeKey) string {
+	if aggregateAttribute.Key != "" && aggregateAttribute.Type == v3.AttributeKeyTypeResource {
+		key := strings.ToLower(aggregateAttribute.Key)
+		return fmt.Sprintf("(simpleJSONHas(lower(labels), '%s') AND lower(labels) like '%%%s%%')", key, key)
+	}
+
+	return ""
+}
+
+func buildResourceSubQuery(bucketStart, bucketEnd int64, fs *v3.FilterSet, groupBy []v3.AttributeKey, aggregateAttribute v3.AttributeKey) (string, error) {
+
+	// BUILD THE WHERE CLAUSE
+	var conditions []string
+	// only add the resource attributes to the filters here
+	rs, err := buildResourceFiltersFromFilterItems(fs)
+	if err != nil {
+		return "", err
+	}
+	conditions = append(conditions, rs...)
+
+	// for aggregate attribute add exists check in resources
+	aggregateAttributeResourceFilter := buildResourceFiltersFromAggregateAttribute(aggregateAttribute)
+	if aggregateAttributeResourceFilter != "" {
+		conditions = append(conditions, aggregateAttributeResourceFilter)
+	}
+
+	groupByResourceFilters := buildResourceFiltersFromGroupBy(groupBy)
+	if len(groupByResourceFilters) > 0 {
+		// TODO: change AND to OR once we know how to solve for group by ( i.e show values if one is not present)
+		groupByStr := "( " + strings.Join(groupByResourceFilters, " AND ") + " )"
+		conditions = append(conditions, groupByStr)
+	}
+	if len(conditions) == 0 {
+		return "", nil
+	}
+	conditionStr := strings.Join(conditions, " AND ")
+
+	// BUILD THE FINAL QUERY
+	query := fmt.Sprintf("(SELECT fingerprint FROM signoz_logs.%s WHERE (seen_at_ts_bucket_start >= %d) AND (seen_at_ts_bucket_start <= %d) AND ", DISTRIBUTED_LOGS_V2_RESOURCE, bucketStart, bucketEnd)
+
+	query = query + conditionStr + ")"
```

Comment:
Please try to avoid this. Why does this only add `)` at the end? Where did the `(` come from?

Reply:
cool, "(" was added in the previous line, but let me clean this.

Reply:
updated